active/0021-rocksdb-message-persistence.md

```diff
@@ -0,0 +1,164 @@
+# An Example of EMQ X Improvement Proposal
+
+## Changelog
+
+* 2022-08-25: @ieQu1 Initial commit
+
+## Abstract
+
+MQTT standard specifies that the MQTT broker must store the messages of the clients that are offline.
+EMQX community edition (up to version 4) fulfilled this promise as long as all broker nodes are online, since the offline messages were stored in RAM.
+Enterprise edition allowed to store the offline messages in a external database.
+
+
+## Motivation
+
+This feature will improve message delivery guarantees of EMQX broker and make it ready for the new markets that require such guarantees.
+
+## Design
+
+### DB schema
+
+The biggest challenge is developing a database schema that enables efficient wildcard topic scans.
+RocksDB supports fast iteration over sorted keys, so naturally we need table keys to be ordered by topic.
+
+We propose to use the following key format:
+
+```
+topic_index-timestamp-clientid
+```
+
+Where `topic_index` is composed of hashes of each topic levels (pseudocode):
+
+```erlang
+index("foo/bar/baz" ++ Rest) ->
+  combine([hash("foo"), hash("bar"), hash("baz"), hash(Rest)]).
+```
+
+Let `N_l` be the maximum number of topic levels that make up the topic index.
+Topic levels deeper than `N_l` should be joined with `/` to a single token before hashing.
+Special care also must be taken for calculating indices of topics with the number of levels less than `N_l`: missing levels should be replaced with values that can't be a part of the topic level, for example character `/`.
+
+In the simplest case `combine` function can just shift bits of each topic level on each iteration, producing a constant-size integer.
+More sophisticated implementations of `combine` function can exist too, for example [Z-order curve](https://en.wikipedia.org/wiki/Z-order_curve) can be used to optimize replaying of topics filters with multiple wildcards following each other (e.g. `foo/?/?`).
+
+With this type of key prefix replaying the topic filters with wildcards can be solved rather elegantly.
+Consider topic filter `foo/?/bar`.
+For the sake of simplicity we assume that `combine` function is a simple bitshift, but it should be noted that any other strategy will work mostly the same.
+Suppose every topic level takes two bits in the resulting hash and `N_l` equals 3.
+Suppose topic index of this topic filter is `101001`.
+We can create a the bitmask corresponding to this topic filter by setting bits corresponding to the wildcard levels to zeros: `110011`.
+
+Therefore iterating the database can be done using the following pseudocode:
+
+```erlang
+replay(TopicFilter) ->
+  Bitmask = gen_bitmask(TopicFilter),
+  TopicIndex = gen_topic_index(TopicFilter) band Bitmask,
+  It = db:seek(TopicIndex),
+  replay(TopicFilter, TopicIndex, Bitmask, It).
+
+replay(TopicFilter, TopicIndex, Bitmask, It) ->
+  {Key, NextIt} = db.next(It),
+  {KTopicIndex, KTimestamp} = parse_key(Key),
+  if TopicIndex bxor (KTopicIndex band Bitmask) =:= 0 ->
```

Comment:
Didn't quite understand this xor operation :see_no_evil: 

Why do we take the xor of those values?  

Apparently, it'll only be `0` when `TopicIndex` has its ones equal to the masked `KTopicIndex` ones.  But, between values of the iterator, won't there be values that give a non-zero xor result? 

For example:

```
...
101001
101010
101011
...
```
?

Reply:
While looking at how RocksDB does the range iterations, it seems that the most it can do is a prefix scan. So, if we want to iterate over the middle segment of the topic as in the example (`a/+/c`), we'd still iterate over all the present third segments.

Maybe we'll have to do the multiple indexing as Ilya has done for retainer to arrange the topics so that a prefix scan can be used?

Reply:
Thanks, it was unnecessary in this case actually. I removed it.